src/base/BuiltIns.ml

```diff
@@ -666,7 +666,7 @@ module ScillaBuiltIns (SR : Rep) (ER : Rep) = struct
     (* Elaborator to run with arbitrary uints *)
     let badd_elab sc ts =
       match ts with
-      | [ s; u ] when s = bnum_typ && is_uint_type u ->
+      | [ PrimType Bnum_typ; PrimType (Uint_typ _) ] ->
```

Comment:
Do we want to duplicate this pattern-match logic everywhere, rather than have the logic centralised in `bnum_typ` and `is_uint_type`? I ask because I have a similar issue in the PR I am working on right now, so if this is our preferred style, then I'll do it the same way there.

Reply:
Right, I was hesitant about this too. Maybe @vaivaswatha can weigh in on this.

Reply:
Wait, with equality this is going to be longer, right? `[ s; u ] when s = bnum_typ && is_uint_type u` would transfer into `[ s; u ] when [%equal: typ] s bnum_typ && is_uint_type u`

Reply:
Well, longer is just a matter of splitting it into multiple lines.

It's more the code duplication I am worried about.

Reply:
I didn't understand this discussion.

Reply:
@vaivaswatha : The point is that `is_uint_type` is defined something like this:

```
let is_uint_type u =
  match u with
  | PrimtType (Uint_typ _) -> true
  | _ -> false
```

By doing what Anton is doing here, we are in essence inlining the pattern-match from `is_uint_type`, which duplicates the code everywhere we do it.

My main concern is that in principle we could change the semantics later on, which would require us to track down every place we have done the inlining, but on the other hand I don't expect this part of the type system to change much.

Reply:
I now understand what you mean. I can't decide what's the right thing though.

Reply:
Ok, I propose to merge this as it is and handle it later. I created an issue for that: https://github.com/Zilliqa/scilla/issues/774